### PR TITLE
Do not wrap CacheFile reentrant r/w locks with ReleasableLock

### PR DESCRIPTION
**Pull request by tlrx**
Today the read/write locks used internally by `CacheFile` object are wrapped into a `ReleasableLock`. This is not strictly required and also prevents usage of the `tryLock()` methods which we would like to use for early releasing of read operations (#58164).

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java [hunk 1]
```diff
@@ -85,7 +85,8 @@ public Path getFile() {
 
     ReleasableLock fileLock() {
         boolean success = false;
-        final ReleasableLock fileLock = readLock.acquire();
+        final ReleasableLock fileLock = new ReleasableLock(readLock);
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:88]: Can we return `fileLock::close` if we make this method return `Releasable` instead?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java [hunk 2]
```diff
@@ -122,9 +124,11 @@ public boolean acquire(final EvictionListener listener) throws IOException {
                     listeners = Collections.unmodifiableSet(newListeners);
                     success = true;
                 } finally {
-                    if (success == false) {
-                        refCounter.decRef();
-                    }
+                    evictionLock.unlock();
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:127]: Before this change we released `evictionLock` after calling `refCounter.decRef()`, but this change reverses the order. Can we keep the order the same?

**Review by DaveCTurner — COMMENTED** [review]
A couple of suggestions/questions.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java]: Sure, in that case we don't need the `ReleasableLock` at all and can return `readLock::unlock`: I pushed ebae53d76e70a2b3282273377e4c85c26f06b301

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java (hunk 2 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java]: Good catch - I've been a bit quick on this. I pushed 2180f565b935dca6ddbbcb55d2abb26a77ba857e

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java [hunk 3]
```diff
@@ -112,20 +112,20 @@ public boolean acquire(final EvictionListener listener) throws IOException {
...
+                final boolean added = newListeners.add(listener);
+                assert added : "listener already exists " + listener;
+                maybeOpenFileChannel(newListeners);
+                listeners = Collections.unmodifiableSet(newListeners);
+                success = true;
+            } finally {
+                if (success == false) {
+                    refCounter.decRef();
                 }
+                evictionLock.unlock();
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:128]: Looks like `refCounter.decRef()` can throw an `UncheckedIOException` so we need another `finally` here.

**Review by DaveCTurner — COMMENTED** [review]
Just one more thing :) Otherwise looks good.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java (hunk 3 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java]: Good catch (bis)

Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java]: I pushed a0e6db7e6a794f38d3da6b85138785d153ebe7a2

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks David, and sorry for the extra iterations :(

@elasticmachine run elasticsearch-ci/docs

**Pull request merged**